File: Algorithm_303_Range Sum Query - Immutable.cpp
#include <iostream>
#include <vector>
using namespace std;
class NumArray {
public:
    NumArray(vector<int> nums) {
        int len=nums.size();
        num.push_back(0);
        for (int i=0;i<len;i++) num.push_back(num[i]+nums[i]);
    }
    
    int sumRange(int i, int j) {
        return num[j+1]-num[i];
    }
private:
    vector<int> num;
}; 

File: Algorithm_303_Range Sum Query - Immutable.java
// 思路：把nums[0:i]（左闭右开）的和存放在sums[i]中，则sumRange(i, j) = sums[j + 1] - sums[i]

class NumArray {
    private int[] sums;
    public NumArray(int[] nums) {
        sums = new int[nums.length + 1];
        for (int i = 1; i <= nums.length; i++) {
            sums[i] = sums[i - 1] + nums[i - 1];
        }
    }
    
    public int sumRange(int i, int j) {
        return sums[j + 1] - sums[i];
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(i,j);
 */